[JIT] Fix double wildcard edge case for may_alias in SchemaInfo and improve formatting

ttanpcs opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81655
* #81653
* #81562
* #81561
* #81575
* #81518
* #81444
* __->__ #81439

- Create c10::AliasTypeSet type def of vector<TypePtr> to match alias_analysis.cpp formatting and improve readability.
- Move canAliasTypeSetsAlias, mapTypeToAliasTypeSet, getAliasTypeSetContainedTypes, and getCorrectList to public in function_schema.h for use in SchemaInfo class. 
**In the future it might be better to find a different home for most of these functions since they don't depend on functionSchema. **
- Created hash function for SchemaArgument
- Add assert to ensure that there is only 1 input and 1 output with each alias set (excluding wildcard)
- Fixed double wildcard input edge case for may_alias. (This is the case where if there is a schema with the form (Tensor(a) a, Tensor(*) b, Tensor(*) c) -> Tensor, and the argument values for 'a' and 'b' cause them to alias, then 'a' may also alias 'c'.
- Added tests for double wildcard case in may_alias, mismatching types in may_alias, and the uniqueness internal assert.

Commit:
[JIT] Fix double wildcard edge case for may_alias in SchemaInfo and improve formatting

[ghstack-poisoned]

Commit:
Update on "[JIT] Fix double wildcard edge case for may_alias in SchemaInfo and improve formatting"




[ghstack-poisoned]

Commit:
Update on "[JIT] Fix double wildcard edge case for may_alias in SchemaInfo and improve formatting"




[ghstack-poisoned]

davidberard98 commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -213,6 +215,9 @@ enum struct TORCH_API SchemaArgType { input, output };
 struct TORCH_API SchemaArgument {
   SchemaArgType type;
   size_t index;
+  bool operator ==(const SchemaArgument& rhs) const {
```
nit: remove the space in `operator ==`

davidberard98 commented on test/cpp/jit/test_schema_info.cpp:
```diff
@@ -173,6 +173,48 @@ TEST(SchemaInfoMayAliasTest, AliasingInputOutput) {
 
+TEST(SchemaInfoMayAliasTest, MultipleWildcardInputs) {
+  SchemaInfo schema(
+      "aten::test.Tensor(Tensor(a) a, Tensor(*) b, Tensor(*) c) -> (Tensor(a), Tensor(*))");
```
mostly just curious, would this handle `fn(Tensor(a -> *) x) -> (Tensor(a), Tensor(*))` correctly?

goldenxuett replied:
I don't think so. I'm not really considering cases where (a -> *) exists but that might be something that I can implement in a later pr.

davidberard98 commented on torch/csrc/utils/schema_info.cpp:
```diff
@@ -191,12 +206,42 @@ std::vector<c10::FunctionSchema> SchemaInfo::getNonDeterministicOps() {
+            } else {
+              for (const auto& set :
+                   arguments_list[i].alias_info()->afterSets()) {
+                TORCH_INTERNAL_ASSERT(
```
is this accurate for returns?

I agree for inputs, it doesn't really make sense to say that two inputs are aliasing (except for in some obscure after-set aliasing where the op would swap data pointers...)

But for outputs, it seems reasonable that you might have two outputs that both alias the same input... for example you might imagine `split_in_half(Tensor(a) x) -> (Tensor(a), Tensor(a))`

either way, can you add some comments on why we have this assertion?

goldenxuett replied:
It's mainly because the check would not be accurate in cases where two outputs both alias the same output and all three are not wildcards. At least from the way I'm currently representing the alias graph, accounting for this case would require going from linear / constant time operations to full dfs searches (up to n^2 time). Furthermore, no ops currently have this aliasing pattern so the assert is to ensure that the class will be accurate in the future (ie: no one runs an op with the above aliasing pattern expecting correct results). 

I'll add a better worded version of the above as a comment in the code.

davidberard98 commented on test/cpp/jit/test_schema_info.cpp:
```diff
@@ -211,5 +253,18 @@ TEST(FunctionSchemaMayContainAliasTest, InputAndOutputContainers) {
+TEST(SchemaInfoInitTest, InvalidSchemaAliases) {
+  SchemaInfo schema(
+      "aten::test.Tensor(Tensor(a) a, Tensor(*) b, Tensor(*) c) -> (Tensor(a), Tensor(*))");
+  ASSERT_THROW(
+      SchemaInfo(
+          "aten::test.Tensor(Tensor(a) a, Tensor(*) b, Tensor(*) c) -> (Tensor(a), Tensor(a))"),
+      c10::Error);
```
^^ this is the test that I think should not fail

Commit:
Update on "[JIT] Fix double wildcard edge case for may_alias in SchemaInfo and improve formatting"


- Create c10::AliasTypeSet type def of vector<TypePtr> to match alias_analysis.cpp formatting and improve readability.
- Move canAliasTypeSetsAlias, mapTypeToAliasTypeSet, getAliasTypeSetContainedTypes, and getCorrectList to public in function_schema.h for use in SchemaInfo class. 
**In the future it might be better to find a different home for most of these functions since they don't depend on functionSchema. **
- Created hash function for SchemaArgument
- Add assert to ensure that there is only 1 input and 1 output with each alias set (excluding wildcard)
- Fixed double wildcard input edge case for may_alias. (This is the case where if there is a schema with the form (Tensor(a) a, Tensor(*) b, Tensor(*) c) -> Tensor, and the argument values for 'a' and 'b' cause them to alias, then 'a' may also alias 'c'.
- Added tests for double wildcard case in may_alias, mismatching types in may_alias, and the uniqueness internal assert.

[ghstack-poisoned]

Commit:
Update on "[JIT] Fix double wildcard edge case for may_alias in SchemaInfo and improve formatting"


- Create c10::AliasTypeSet type def of vector<TypePtr> to match alias_analysis.cpp formatting and improve readability.
- Move canAliasTypeSetsAlias, mapTypeToAliasTypeSet, getAliasTypeSetContainedTypes, and getCorrectList to public in function_schema.h for use in SchemaInfo class. 
**In the future it might be better to find a different home for most of these functions since they don't depend on functionSchema. **
- Created hash function for SchemaArgument
- Add assert to ensure that there is only 1 input and 1 output with each alias set (excluding wildcard)
- Fixed double wildcard input edge case for may_alias. (This is the case where if there is a schema with the form (Tensor(a) a, Tensor(*) b, Tensor(*) c) -> Tensor, and the argument values for 'a' and 'b' cause them to alias, then 'a' may also alias 'c'.
- Added tests for double wildcard case in may_alias, mismatching types in may_alias, and the uniqueness internal assert.

[ghstack-poisoned]

Gamrix commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -602,4 +607,15 @@ inline std::string toString(const FunctionSchema& schema) {
+  struct hash<c10::SchemaArgument> {
+    size_t operator()(const c10::SchemaArgument& arg) const
+    {
+      return std::hash<std::string>()(std::to_string((int)(arg.type)) + ":" + std::to_string(arg.index));
```
Hash the values directly instead of converting the values to strings.

goldenxuett replied:
Wouldn't I need to combine the hashes then? I thought that converting the values to a string would be simpler then messing with combining the hashes of type and index, but I might be mistaken about this.

davidberard98 replied:
https://github.com/pytorch/pytorch/blob/master/c10/util/hash.h#L39-L41

Gamrix commented on test/cpp/jit/test_schema_info.cpp:
```diff
@@ -173,6 +173,48 @@ TEST(SchemaInfoMayAliasTest, AliasingInputOutput) {
+  SchemaInfo schema(
+      "aten::test.Tensor(Tensor(a) a, Tensor(*) b, Tensor(*) c) -> (Tensor(a), Tensor(*))");
+  ASSERT_TRUE(schema.may_alias(
+      {c10::SchemaArgType::input, 0}, {c10::SchemaArgType::output, 0}));
```
a `namespace ArgType = c10::SchemaArgType;` declaration would make ti so that you could write this test much more readable.

Gamrix commented on torch/csrc/utils/schema_info.cpp:
```diff
@@ -208,15 +259,26 @@ void SchemaInfo::generateAliasMaps() {
+          if (wildcard_set_.count({c10::SchemaArgType::input, i})) {
+            wildcard_set_.insert({c10::SchemaArgType::input, j});
+          }
+          if (wildcard_set_.count({c10::SchemaArgType::input, j})) {
+            wildcard_set_.insert({c10::SchemaArgType::input, i});
```
change to `elseif`

Commit:
Update on "[JIT] Fix double wildcard edge case for may_alias in SchemaInfo and improve formatting"


- Create c10::AliasTypeSet type def of vector<TypePtr> to match alias_analysis.cpp formatting and improve readability.
- Move canAliasTypeSetsAlias, mapTypeToAliasTypeSet, getAliasTypeSetContainedTypes, and getCorrectList to public in function_schema.h for use in SchemaInfo class. 
**In the future it might be better to find a different home for most of these functions since they don't depend on functionSchema. **
- Created hash function for SchemaArgument
- Add assert to ensure that there is only 1 input and 1 output with each alias set (excluding wildcard)
- Fixed double wildcard input edge case for may_alias. (This is the case where if there is a schema with the form (Tensor(a) a, Tensor(*) b, Tensor(*) c) -> Tensor, and the argument values for 'a' and 'b' cause them to alias, then 'a' may also alias 'c'.
- Added tests for double wildcard case in may_alias, mismatching types in may_alias, and the uniqueness internal assert.

[ghstack-poisoned]

Gamrix commented on torch/csrc/utils/schema_info.cpp:
```diff
@@ -191,12 +206,66 @@ std::vector<c10::FunctionSchema> SchemaInfo::getNonDeterministicOps() {
+        if (arguments_list[i].alias_info()->isWildcardAfter()) {
+          wildcard_set_.insert({type, i});
+        } else {
+          // This check is to ensure that the FunctionSchema will accurately
+          // be represented when calling may_alias and may_contain_alias
```
Please add in the comment what is the error case we are trying to avoid.

Commit:
Update on "[JIT] Fix double wildcard edge case for may_alias in SchemaInfo and improve formatting"


- Create c10::AliasTypeSet type def of vector<TypePtr> to match alias_analysis.cpp formatting and improve readability.
- Move canAliasTypeSetsAlias, mapTypeToAliasTypeSet, getAliasTypeSetContainedTypes, and getCorrectList to public in function_schema.h for use in SchemaInfo class. 
**In the future it might be better to find a different home for most of these functions since they don't depend on functionSchema. **
- Created hash function for SchemaArgument
- Add assert to ensure that there is only 1 input and 1 output with each alias set (excluding wildcard)
- Fixed double wildcard input edge case for may_alias. (This is the case where if there is a schema with the form (Tensor(a) a, Tensor(*) b, Tensor(*) c) -> Tensor, and the argument values for 'a' and 'b' cause them to alias, then 'a' may also alias 'c'.
- Added tests for double wildcard case in may_alias, mismatching types in may_alias, and the uniqueness internal assert.

[ghstack-poisoned]

davidberard98 reviewed (approved):
looks good!!
